File: server/src/main/java/com/ahmadramadhan/mudahtitip/consignment/dto/ConsignmentRequest.java
package com.ahmadramadhan.mudahtitip.consignment.dto;

import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Positive;
import jakarta.validation.constraints.PositiveOrZero;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDate;

/**
 * DTO for creating a consignment.
 */
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class ConsignmentRequest {

    @NotNull(message = "Product ID wajib diisi")
    private Long productId;

    @NotNull(message = "Shop ID wajib diisi")
    private Long shopId;

    @NotNull(message = "Jumlah wajib diisi")
    @Positive(message = "Jumlah harus positif")
    private Integer quantity;

    @NotNull(message = "Harga jual wajib diisi")
    @Positive(message = "Harga harus positif")
    private BigDecimal sellingPrice;

    @PositiveOrZero(message = "Komisi tidak boleh negatif")
    private BigDecimal commissionPercent;

    private LocalDate consignmentDate;

    private LocalDate expiryDate;

    private String notes;
}

File: client/lib/features/consignment/data/models/consignment_request.dart
import 'package:equatable/equatable.dart';

/// Request model for creating a consignment.
class ConsignmentRequest extends Equatable {
  final int productId;
  final int shopId;
  final int quantity;
  final double sellingPrice;
  final double commissionPercent;
  final DateTime? consignmentDate;
  final DateTime? expiryDate;
  final String? notes;

  const ConsignmentRequest({
    required this.productId,
    required this.shopId,
    required this.quantity,
    required this.sellingPrice,
    this.commissionPercent = 0.0, // Default to 0 for shop owner flow
    this.consignmentDate,
    this.expiryDate,
    this.notes,
  });

  Map<String, dynamic> toJson() {
    return {
      'productId': productId,
      'shopId': shopId,
      'quantity': quantity,
      'sellingPrice': sellingPrice,
      'commissionPercent': commissionPercent,
      if (consignmentDate != null)
        'consignmentDate': consignmentDate!.toIso8601String().split('T')[0],
      if (expiryDate != null)
        'expiryDate': expiryDate!.toIso8601String().split('T')[0],
      if (notes != null) 'notes': notes,
    };
  }

  @override
  List<Object?> get props => [
    productId,
    shopId,
    quantity,
    sellingPrice,
    commissionPercent,
    consignmentDate,
    expiryDate,
    notes,
  ];
}
